compiler/rustc_hir_typeck/src/fn_ctxt/mod.rs

```diff
@@ -49,6 +49,13 @@ pub struct FnCtxt<'a, 'tcx> {
     /// eventually).
     pub(super) param_env: ty::ParamEnv<'tcx>,
 
+    /// The `DefId` of the item in whose context we are performing inference or typeck.
+    /// It is used to check whether an opaque type use is a defining use.
+    ///
+    /// Its default value is `DefiningAnchor::Error`, this way it is easier to catch errors that
+    /// might come up during inference or typeck.
+    pub(super) defining_use_anchor: DefiningAnchor,
```

Comment:
Can you instead move this to `Inherited`. Considering that the we only use the `FnCtxt` during hir typeck, maybe the `defining_use_anchor` should only be a `DefId` as we should always use `DefiningAnchor::Bind(def_id_of_named_function)`

Reply:
mostly done that. Making it just be the `DefId` is a bit annoying as now all use sites need to `DefiningAnchor::Bind(self.defining_use_anchor)`. I want to see if I can simplify that with some `impl Into<DefiningAnchor>` impls, but this PR is gonna be bitrotty enough on its own.